lib/puma/binder.rb

```diff
@@ -50,12 +50,9 @@ def env(sock)
 
     def close
       @ios.each { |i| i.close }
-      @unix_paths.each do |i|
-        # Errno::ENOENT is intermittently raised
-        begin
-          unix_socket = UNIXSocket.new i
-          unix_socket.close
-        rescue Errno::ENOENT
+      @unix_paths.uniq.each do |i|
+        if i && !File.socket?(i) && File.exist?(i)
```

Comment:
Why not check existence before checking if it's a socket?

Reply:
I believe File.socket?(i) is the check for whether it should be left in place due to a restart.

I left `File.exist?(i)` in just so an error wouldn't be generated if something like a graceful shutdown was run and the 'other side' removed the file?

Reply:
Ohhhh I get it. So on a restart, the file descriptor is not a socket anymore (from Ruby's perspective)?

Reply:
Umm, backwards maybe?  On restart, we want to leave connections intact, so we don't want to unlink the file?

So, on restart, since the 'other side' still has the socket open, `File.socket?(i)` is true, but it's negated in the logic to unlink the file...

Reply:
I don't see how this logic makes sense at all. Why would the file exist and be a regular file? File.socket? will return true even if there is nothing connected to it I believe.

Reply:
> File.socket? will return true even if there is nothing connected to it I believe

Quick check on Travis showed that `File.socket? __FILE__` returned false.

>  Why would the file exist and be a regular file? 

Normally it wouldn't.  But if `File.socket?` does work, and hence, nothing is connected, shouldn't it be unlinked?

Reply:
So you're understanding is that `File.socket?` will return false if nothing is connected? I've never heard of that behavior and kind of doubt it because the entry type is backed into the directory entry from the beginning.

Reply:
As I stated above, I ran it thru several os's and Ruby versions on Travis
```ruby
puts "File.socket? __FILE__  #{File.socket? __FILE__}", RUBY_DESCRIPTION
```

all print the first line as:
```
File.socket? __FILE__  false
```

See https://travis-ci.org/MSP-Greg/testing-travis/builds/561254063

I'm not sure if your implying that once something is connected, and then removed or closed, `File.socket?` will return true?  If so, I didn't check for that.

Reply:
I'm unsure what you're testing with that code because of course a `.rb` file is not a socket. My point is that once a unix domain socket is created, it will return true from `File.socket?` no matter if it's used or not.

So that would mean your code is assuming that maybe the path specified to be a socket is for some reason not actually a domain socket but does exist? I don't see how that would be the case.

Reply:
Ok.  As mentioned above, until a recent commit (but before this PR), the code included
```ruby
@unix_paths.each { |i| File.unlink i }
```

So every file contained in @unix_paths was unlinked.  Also note that there have been and continue to be a number of `Errno::ECONNREFUSED` errors raised in Travis testing.  Some of the issues related to UNIXSockets were related to connection issues on restart.

Hence, I wasn't sure if all the items in @unix_paths should be affected, which resulted in the conditional.  Since I'm a windows guy, I also checked the Ruby UNIXSocket test files, and they were using `File.socket?` in places.

After those changes, all the tests pass, and the `Errno::ECONNREFUSED` errors disappeared.   The other changes were just things I thought stabilized testing.

Also, note the last commit where I returned a line in `inherit_unix_listener` to the original code:
```
if fd.kind_of? TCPServer
```

This seems very odd.

I don't recall every detail of this, I can review it more tomorrow.